File: BITEC/current lesson/oop/src/FreitagNachmittag/Hamster.java
package FreitagNachmittag;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Hamster {
    // Felder
    private int x;
    private int y;
    private String darstellung;
    private boolean hatHunger;

    // (hat) Beziehungen
    private Spielfeld spielfeld;
    private List<Samen> backenSpeicher;

    // Konstruktor
    public Hamster(Spielfeld spielfeld) {
        hatHunger = false;
        darstellung = "🐹";

        backenSpeicher = new ArrayList<>();
        this.spielfeld = spielfeld;

        spielfeld.weiseHamsterZu(this);
        platziereMichAufFeld();
    }

    private void platziereMichAufFeld() {
        Random random = new Random();
        // TODO: Frage ab, ob im Spielfeld ein Bodensymbol auf den Koordinaten x und y ist.
        //  Wenn ja, setze den Hamster dort hin.
        //  Wenn nein, probier es nochmal.
        //  Konzepte: while und if (oder denke an ein do-while für kürzeren Code.)
//        boolean platziert = false;
//        while (!platziert) {
//            x = random.nextInt(spielfeld.getGroesse());
//            y = random.nextInt(spielfeld.getGroesse());
//
//            if (spielfeld.getSpielfeld()[y][x].equals(spielfeld.getBodenSymbol())) {
//               platziert = true;
//            }
//        }

        // aber vl. besser das hier:
        var bodensymbol = spielfeld.getBodenSymbol();
        var feld = spielfeld.getSpielfeld();

        do {
            x = random.nextInt(spielfeld.getGroesse());
            y = random.nextInt(spielfeld.getGroesse());
        } while (!feld[y][x].equals(bodensymbol));

        spielfeld.getSpielfeld()[y][x] = darstellung;
    }

    // Methoden
    public void bewegen() {
        Random random = new Random();
        int index = random.nextInt(0, Richtung.values().length);
        Richtung richtung = Richtung.values()[index];

        spielfeld.bewegeHamster(this, richtung);
    }

    public void metabolismn() {
        // werde zufällig hungrig

        // stehe ich auf einem Feld mit essen?
        // wenn ja, dann rufe methode essen auf
        // ansonsten rufe methode hamstern auf
    }

    public void essen() {
        // hamster wird nicht mehr hungrig.
        hatHunger = false;
        // hamster sagt dem spielfeld, der samen ist weg
        spielfeld.hamsterIsstSamen(this);
    }

    public void hamstern() {
        // hamster merkt sich, dass ein neues Samen Objekt gespeichtert wird.
        // hamster sagt dem spielfeld, der samen ist weg
        spielfeld.hamsterHamstertSamen(this);
    }

    @Override
    public String toString() {
        return darstellung;
    }

    // get-set Methoden
    public int getX() {
        return x;
    }

    public void setX(int x) {
        this.x = x;
    }

    public int getY() {
        return y;
    }

    public void setY(int y) {
        this.y = y;
    }

    public String getDarstellung() {
        return darstellung;
    }

    public void setDarstellung(String darstellung) {
        this.darstellung = darstellung;
    }

    public boolean isHatHunger() {
        return hatHunger;
    }

    public void setHatHunger(boolean hatHunger) {
        this.hatHunger = hatHunger;
    }

    public Spielfeld getSpielfeld() {
        return spielfeld;
    }

    public void setSpielfeld(Spielfeld spielfeld) {
        this.spielfeld = spielfeld;
    }

    public List<Samen> getBackenSpeicher() {
        return backenSpeicher;
    }

    public void setBackenSpeicher(List<Samen> backenSpeicher) {
        this.backenSpeicher = backenSpeicher;
    }
}

File: JET/modul_1_c#_basics/L04KlassenMethoden/Exercise2/Hamster.cs
namespace Hamster;

public class Hamster
{
    // Felder
    private (int x, int y) position;
    private String representation;
    private static String hungryRepresentation = "😡";
    private static String fedRepresentation = "🐹";
    private bool isHungry;

    // (hat) Beziehungen
    private Plane plane;
    private List<Seed> mouth = new List<Seed>();

    // Konstruktor
    public Hamster(Plane plane)
    {
        isHungry = false;
        representation = fedRepresentation;
        this.plane = plane;

        PositionAndManageHamster();
    }

    // Methoden
    private void PositionAndManageHamster()
    {
        var random = new Random();
        bool done;
        int x, y;

        do
        {
            x = random.Next(plane.GetSize());
            y = random.Next(plane.GetSize());

            done = plane.AssignInitialPosition(this, (x, y));
        } while (!done);

        position = (x, y);
    }

    public void Move()
    {
        var random = new Random();
        int index = random.Next(Enum.GetValues<Direction>().Length);
        var direction = Enum.GetValues<Direction>()[index];

        plane.Position(this, direction);
    }

    public void NutritionBehaviour()
    {
        // werde zufällig hungrig
        var random = new Random();

        if (random.NextDouble() < 0.1)
        {
            isHungry = true;
            representation = hungryRepresentation;
        }

        // ist ein Samen unter mir (hamster) ?
        if (plane.ContainsSeed(position))
        {
            // habe ich hunger?
            if (isHungry)
            {
                // wenn ja, dann rufe methode essen auf
                EatSeedFromTile();
            }
            else
            {
                // ansonsten rufe methode hamstern auf
                PutInMouth();
            }

        }
        else
        {
            if (isHungry && mouth.Any())
            {
                EatSeedFromMouth();
            }
        }
    }

    private void EatSeedFromMouth()
    {
        // hamster wird nicht mehr hungrig.
        Eat();

        // hamster entfernt den Samen aus dem Mund - könnte Queue sein, statt List. Hier haben wir verhalten eines Stacks.
        mouth.RemoveAt(0);
    }

    public void EatSeedFromTile()
    {
        // hamster wird nicht mehr hungrig.
        Eat();

        // hamster sagt dem spielfeld, der samen ist weg
        plane.HamsterIsEatingSeeds(this);
    }

    private void Eat()
    {
        isHungry = false;
        representation = fedRepresentation;
    }

    public void PutInMouth()
    {
        // hamster merkt sich, dass ein neues Samen Objekt gespeichtert wird.
        var samen = plane.GetSamen(position);
        mouth.Add(samen);

        // hamster sagt dem spielfeld, der samen ist weg
        plane.HamsterIsStoringSeeds(this);
    }

    public override String ToString()
    {
        return representation;
    }

    // get-set Methoden
    public (int x, int y) GetPosition()
    {
        return position;
    }

    public void SetPosition(int x, int y)
    {
        position.x = x;
        position.y = y;
    }

    public void SetPosition((int x, int y) position)
    {
        this.position = position;
    }

    public int GetX()
    {
        return position.x;
    }

    public void SetX(int x)
    {
        position.x = x;
    }

    public int GetY()
    {
        return position.y;
    }

    public void SetY(int y)
    {
        position.y = y;
    }

    public void SetY((int x, int y) position)
    {
        this.position = position;
    }

    public String GetRepresentation()
    {
        return representation;
    }

    public static String GetHungryRepresentation()
    {
        return hungryRepresentation;
    }

    public static String GetFedRepresentation()
    {
        return fedRepresentation;
    }
}
